src/Models/HydrostaticFreeSurfaceModels/update_hydrostatic_free_surface_model_state.jl

```diff
@@ -28,8 +28,380 @@ they are called in the end.
+                field_2[region][Nc+1:Nc+Hc, 1:Nc, k]   .=         field_2[region_E][1:Hc, 1:Nc, k]
+                field_2[region][Nc+1:Nc+Hc, Nc+1, k]   .=         field_2[region_N][1:Hc, 1, k]
+                field_2[region][1-Hc:0, 2:Nc+1, k]     .= reverse(field_1[region_W][1:Nc, Nc+1-Hc:Nc, k], dims=1)' * plmn
+                field_2[region][1-Hc:0, 1, k]          .=         field_1[region_S][1, Nc+1-Hc:Nc, k] * plmn
+                #- N + S Halo for field_2:
+                field_2[region][2:Nc, Nc+1:Nc+Hc, k]   .= reverse(field_1[region_N][1:Hc, 2:Nc, k], dims=2)'
+                if Hc > 1
+                    field_2[region][1, Nc+2:Nc+Hc, k]   = reverse(field_2[region_W][1, Nc+2-Hc:Nc, k]) * plmn
+                end
+                field_2[region][1:Nc, 1-Hc:0, k]       .=         field_2[region_S][1:Nc, Nc+1-Hc:Nc, k]
+            end
+        else
+            #- even face number (2,4,6):
+            region_E = mod(region + 1, 6) + 1
+            region_N = mod(region + 0, 6) + 1
+            region_W = mod(region + 4, 6) + 1
+            region_S = mod(region + 3, 6) + 1
+            for k in -Hz+1:Nz+Hz
+                #- E + W Halo for field_1:
+                field_1[region][Nc+1:Nc+Hc, 2:Nc, k]   .= reverse(field_2[region_E][2:Nc, 1:Hc, k], dims=1)'
+                if Hc > 1
+                    field_1[region][Nc+2:Nc+Hc, 1, k]  .= reverse(field_1[region_S][Nc+2-Hc:Nc, 1, k]) * plmn
+                end
+                field_1[region][1-Hc:0, 1:Nc, k]       .=         field_1[region_W][Nc+1-Hc:Nc, 1:Nc, k]
+                #- N + S Halo for field_1:
+                field_1[region][1:Nc, Nc+1:Nc+Hc, k]   .=         field_1[region_N][1:Nc, 1:Hc, k]
+                field_1[region][Nc+1, Nc+1:Nc+Hc, k]    =         field_1[region_E][1, 1:Hc, k]
+                field_1[region][2:Nc+1, 1-Hc:0, k]     .= reverse(field_2[region_S][Nc+1-Hc:Nc, 1:Nc, k], dims=2)' * plmn
+                field_1[region][1, 1-Hc:0, k]           =         field_2[region_W][Nc+1-Hc:Nc, 1, k] * plmn
+                #- E + W Halo for field_2:
+                field_2[region][Nc+1:Nc+Hc, 2:Nc+1, k] .= reverse(field_1[region_E][1:Nc, 1:Hc, k], dims=1)' * plmn
+                if Hc > 1
+                    field_2[region][Nc+2:Nc+Hc, 1, k]  .= reverse(field_2[region_S][Nc+2-Hc:Nc, 1, k]) * plmn
+                end
+                field_2[region][1-Hc:0, 1:Nc, k]       .=         field_2[region_W][Nc+1-Hc:Nc, 1:Nc, k]
+                field_2[region][1-Hc:0, Nc+1, k]       .= reverse(field_1[region_N][1, 2:Hc+1, k])
+                #- N + S Halo for field_2:
+                field_2[region][1:Nc, Nc+1:Nc+Hc, k]   .=         field_2[region_N][1:Nc, 1:Hc, k]
+                field_2[region][2:Nc, 1-Hc:0, k]       .= reverse(field_1[region_S][Nc+1-Hc:Nc, 2:Nc, k], dims=2)'
+                field_2[region][1, 1-Hc:0, k]           =         field_1[region_W][Nc+1-Hc:Nc, 1, k]
+            end
+        end
+
+    end
+
+    return nothing
+end
```

Comment:
These methods don't seem to belong in `HydrostaticFreeSurfaceModel`.
Judging from their name they belong either in BoundaryConditions or in MultiRegion? I presume they are only applicable for CubedSphereFields but I don't see a typed argument to ensure this.

Also, I need to say that their structure reads more like Fortran and not Julia: there are a lot of `if` statements. Also they miss `@inbounds` when array slicing occurs.

Reply:
I suggest something like:

```Julia
fill_cubed_sphere_halo_regions!(fields::Tuple, locations::Tuple{<:Face, <:Face}, locations::Tuple{<:Face, <:Face}, signed=true) =
    fill_cubed_sphere_halo_regions!(fields[1], fields[2], location, locations; signed)

function fill_cubed_sphere_halo_regions!(field1::CubedSphereField, field2::CubedSphereField, ::Tuple{<:Face, <:Face}, ::Tuple{<:Face, <:Face}, signed=true)
...
end
```

Reply:
> These methods don't seem to belong in `HydrostaticFreeSurfaceModel`. Judging from their name they belong either in BoundaryConditions or in MultiRegion? I presume they are only applicable for CubedSphereFields but I don't see a typed argument to ensure this.
> 
> Also, I need to say that their structure reads more like Fortran and not Julia: there are a lot of `if` statements. Also they miss `@inbounds` when array slicing occurs.
> 
> `reverse` also allocates arrays -- this will create huge problems on GPUs but that can be left for later

@navidcy, I agree that these methods only apply to multi-region cubed sphere grids, and should be relocated to a script in src/MultiRegion. Regarding the code structure, please examine the commit [11ae8a3](https://github.com/CliMA/Oceananigans.jl/commit/11ae8a37d99b441da3dc1d6c0235678779b68e0c) in the branch [sb-ss/cubed-sphere-kernels](https://github.com/CliMA/Oceananigans.jl/tree/sb-ss/cubed-sphere-kernels), where I am enhancing these methods by integrating kernel functions, enabling their execution on a single GPU. I plan to submit a pull request for this branch by tomorrow.